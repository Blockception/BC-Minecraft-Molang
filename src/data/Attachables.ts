import { Data } from "./Data";

/**Molang data for animations controllers*/
export namespace Attachables {
  /**The list of usable variables for this specific type*/
  export const Variables: Data[] = [{id: "is_enchanted"}];
  /**The list of specific contexts usable for this specific type*/
  export const Contexts: Data[] = [{ id: "is_first_person" }, { id: "item_slot" }];
  /**The list of usable temp variables for this specific type*/
  export const Temps: Data[] = [];
}
