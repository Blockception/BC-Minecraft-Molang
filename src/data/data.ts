import { Types } from "bc-minecraft-bedrock-types";

/**The base type of a molang data item*/
export type Data = Types.Identifiable & Types.Documentated;
