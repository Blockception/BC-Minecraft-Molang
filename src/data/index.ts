/*	Auto generated	*/
export * from "./animations";
export * from "./animations-controllers";
export * from "./attachables";
export * from "./blocks";
export * from "./data";
export * from "./entities";
export * from "./features-rules";
export * from "./general";
export * from "./items";
export * from "./molang-data";
export * from "./molang-function";
export * from "./particles";
export * from "./render-controllers";

