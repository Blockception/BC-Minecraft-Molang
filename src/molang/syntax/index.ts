export * from './builder'
export * from './errors'
export * from './nodes'
export * from './operators'
export * from './parse'
export * from './tokens'
export * from './util'