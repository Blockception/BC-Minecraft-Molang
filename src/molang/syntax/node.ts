import { LexicalNode } from "./lexical/lexical";

export type Syntax =
  | AccessOperationNode
  | AssignmentNode
  | CompareNode
  | ConstantNode
  | OperationNode
  | FunctionCallNode
  | UnknownNode;

export enum Node {
  unknown,
  access,
  assignment,
  constant,
  function,
  identifier,
  operation,
  compare,
}

export interface Expression {
  nodes: Syntax;
}

export interface AssignmentNode {
  type: Node.assignment;
  receiver: Syntax;
  value: Syntax;
}

export namespace AssignmentNode {
  /**
   *
   * @param accessor the node with . or ->
   * @param base
   * @param property
   * @returns
   */
  export function create(receiver: Syntax, value: Syntax): AssignmentNode {
    return { type: Node.assignment, receiver, value };
  }
}

export interface AccessOperationNode {
  type: Node.access;
  accessor: LexicalNode; // -> or .
  base: Syntax;
  property: Syntax;
}

export namespace AccessOperationNode {
  /**
   *
   * @param accessor the node with . or ->
   * @param base
   * @param property
   * @returns
   */
  export function create(accessor: LexicalNode, base: Syntax, property: Syntax): AccessOperationNode {
    return { type: Node.access, accessor, base, property };
  }
}

export interface OperationNode {
  type: Node.operation;
  operation: LexicalNode;
  parameters: Syntax[];
}

export namespace OperationNode {
  export function create(operation: LexicalNode, ...parameters: Syntax[]): OperationNode {
    return { type: Node.operation, operation, parameters };
  }
}

export interface CompareNode {
  type: Node.compare;
  a: Syntax;
  b: Syntax;
  operator: LexicalNode;
}

export namespace CompareNode {
  export function create(a: Syntax, b: Syntax, operator: LexicalNode): CompareNode {
    return { type: Node.compare, a, b, operator };
  }
}

export interface UnknownNode {
  type: Node.unknown;
  base: LexicalNode;
}

export namespace UnknownNode {
  export function create(base: LexicalNode): UnknownNode {
    return { type: Node.unknown, base };
  }
}

export interface ConstantNode {
  type: Node.constant | Node.identifier;
  base: LexicalNode;
}

export namespace ConstantNode {
  export function create(base: LexicalNode, type: Node.constant | Node.identifier): ConstantNode {
    return { type, base };
  }
}

export interface FunctionCallNode {
  type: Node.function;
  identifier: Syntax;
  parameters: Syntax[];
}

export namespace FunctionCallNode {
  export function create(identifier: Syntax, parameters: Syntax[]): FunctionCallNode {
    return { type: Node.function, identifier, parameters };
  }
}

export function convertToString(node: Syntax): string {
  switch (node.type) {
    case Node.compare:
      return `${convertToString(node.a)}${node.operator.text}${convertToString(node.b)}`;
    case Node.identifier:
    case Node.constant:
    case Node.unknown:
      return node.base.text;
    case Node.operation:
      return `(${node.parameters.map(convertToString).join(node.operation.text)})`;
    case Node.access:
      return `${convertToString(node.base)}${node.accessor.text}${convertToString(node.property)}`;
    case Node.function:
      return `${convertToString(node.identifier)}(${node.parameters.map(convertToString).join(", ")})`;
    case Node.assignment:
      return `${convertToString(node.receiver)}=${convertToString(node.value)}`;
  }
}
