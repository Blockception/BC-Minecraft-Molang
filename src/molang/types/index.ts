/*	Auto generated	*/

export * from "./context";
export * from "./geometries";
export * from "./materials";
export * from "./math";
export * from "./queries";
export * from "./temps";
export * from "./textures";
export * from "./variables";