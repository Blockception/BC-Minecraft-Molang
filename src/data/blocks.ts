import { Data } from "./data";

/**Molang data for blocks*/
export namespace Blocks {
  /**The list of usable variables for this specific type*/
  export const Variables: Data[] = [];
  /**The list of specific contexts usable for this specific type*/
  export const Contexts: Data[] = [];
    /**The list of usable temp variables for this specific type*/
  export const Temps: Data[] = [];
}
