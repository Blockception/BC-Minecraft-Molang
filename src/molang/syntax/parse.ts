import { OffsetWord } from "bc-minecraft-bedrock-types/lib/types";
import { toBlocks } from "./lexical/blocks";
import { parse } from "./lexical/parsing";
import { buildTree } from "./tree-builder";

export function parseMolang(line: OffsetWord) {
  const tokens = parse(line);
  const blocks = toBlocks(tokens);
  return blocks.map(buildTree);
}
