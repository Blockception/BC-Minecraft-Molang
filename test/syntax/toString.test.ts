import { OffsetWord } from "bc-minecraft-bedrock-types/lib/types";
import { parseMolang } from "../../src/molang/syntax/parse";
import { convertToString } from "../../src/molang/syntax/node";
import { valid_syntaxes } from "../data/dataset";

describe("molang - syntax", () => {
  describe("should be able to parse and match the syntax tree generated", () => {
    test.each(valid_syntaxes.filter((s) => s.includes("&&")))("%#. %s", (s) => {
      const statements = parseMolang(OffsetWord.create(s, 0));

      statements.forEach((n) => {
        expect(convertToString(n)).toMatchSnapshot();
      });
    });
  });
});
