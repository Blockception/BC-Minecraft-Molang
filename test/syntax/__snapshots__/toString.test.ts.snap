// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`molang - syntax should be able to parse and match the syntax tree generated 0. !variable.swaying && query.all_animations_finished 1`] = `"((variable.variable.swaying&&query.query.all_animations_finished))"`;

exports[`molang - syntax should be able to parse and match the syntax tree generated 2. variable.walking = query.modified_move_speed > 0.01 && !variable.is_rolled_up; 1`] = `"variable.variable.walking=(query.query.modified_move_speed>0.01&&(variable.variable.is_rolled_up))"`;
